chrome/content/api/FPVS/implementation.js

```diff
@@ -0,0 +1,394 @@
+/* eslint-disable object-shorthand */
+
+"use strict";
+
+(function (exports) {
+
+  var { ExtensionCommon } = ChromeUtils.import(
+    "resource://gre/modules/ExtensionCommon.jsm"
+  );
+  var { ExtensionSupport } = ChromeUtils.import(
+    "resource:///modules/ExtensionSupport.jsm"
+  );
+  var { MailServices } = ChromeUtils.import(
+    "resource:///modules/MailServices.jsm"
+  );
+  var { Services } = ChromeUtils.import("resource://gre/modules/Services.jsm");
+
+  let windowListener;
+  const logEnabled = true;
+
+  function log(...a) {
+    if (logEnabled) {
+      console.log("FPVS Experiment", ...a);
+    }
+  }
+
+  class WindowListener extends ExtensionCommon.EventEmitter {
+    constructor(extension) {
+      super();
+      this.extension = extension;
+      this.callbackCount = 0;
+      this.currentFolder = new Map();
+    }
+
+    get listenerId() {
+      return `experiment_listener_${this.extension.uuid}_${this.extension.instanceId}`;
+    }
+
+    // Implements nsIMsgFolderListener
+    msgsMoveCopyCompleted(
+      aMove,
+      aSrcMsgs,
+      aDestFolder,
+      aDestMsgs
+    ) {
+      // We have to assume, that the folder action was on the active window.
+      let mail3Pane = Services.wm.getMostRecentWindow("mail:3pane");
+
+      if (aDestFolder == this.currentFolder.get(mail3Pane)) {
+        // Still remotely possible that someone else could be copying
+        // into the same folder at the same time as us, but this is
+        // the best we can do until they fix the event bug.
+        //  FolderPaneSwitcher.onDragDrop({ type: "msgsMoveCopyCompleted" });
+        windowListener.emit("dragdrop", mail3Pane, "folderListener", "msgsMoveCopyCompleted");
+      }
+    }
+    folderMoveCopyCompleted(aMove, aSrcFolder, aDestFolder) {
+      if (aDestFolder == this.currentFolder.get(mail3Pane)) {
+        // Still remotely possible that someone else could be copying
+        // into the same folder at the same time as us, but this is
+        // the best we can do until they fix the event bug.
+        //         FolderPaneSwitcher.onDragDrop({ type: "folderMoveCopyCompleted" });
+        windowListener.emit("dragdrop", mail3Pane, "folderListener", "folderMoveCopyCompleted");
+      }
+    }
+
+    handleEvent(event) {
+      // This function handles the events received from the window and forwards
+      // them to the WebExtension by emitting a custom event (this is an EventEmitter)
+      // with the information/parameters we want to send to the WebExtension event
+      // listeners.
+      let win = event.target.ownerGlobal;
+      log(event.type);
```

Comment:
Do we need this debugging statement?

Reply:
Setting `logEnabled = false` to disable all logs.